### PR TITLE
Run node code in virtualenv

### PR DESCRIPTION
**Pull request by sean-smith**
This solves https://github.com/aws/aws-parallelcluster/issues/1157

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: recipes/_setup_python.rb [hunk 1]
```diff
@@ -53,4 +64,7 @@
     # FIXME: https://github.com/poise/poise-python/issues/133
     pip_version '18.0'
   end
+  python_virtualenv node['cfncluster']['virtualenv'] do
+      pip_version '18.0'
+  end
```
Comment (lukeseawalker) [inline recipes/_setup_python.rb]: Can we clean up all this logic and just create a python virtual env valid for all the OSs?

Comment (sean-smith) [reply recipes/_setup_python.rb]: done

File: attributes/default.rb [hunk 2]
```diff
@@ -18,10 +18,12 @@
 default['cfncluster']['sources_dir'] = "#{node['cfncluster']['base_dir']}/sources"
 default['cfncluster']['scripts_dir'] = "#{node['cfncluster']['base_dir']}/scripts"
 default['cfncluster']['license_dir'] = "#{node['cfncluster']['base_dir']}/licenses"
+default['cfncluster']['virtualenv'] = "#{node['cfncluster']['base_dir']}/virtualenv"
+# Python version
+default['cfncluster']['python-version'] = '2.7.8'
```
Comment (lukeseawalker) [inline attributes/default.rb]: why this version?

File: attributes/default.rb [hunk 3]
```diff
@@ -108,8 +110,9 @@
     default['cfncluster']['base_packages'] = %w[vim ksh tcsh zsh openssl-devel ncurses-devel pam-devel net-tools openmotif-devel
                                                 libXmu-devel hwloc-devel db4-devel tcl-devel automake autoconf pyparted libtool
                                                 httpd boost-devel redhat-lsb mlocate mpich-devel R atlas-devel fftw-devel
-                                                libffi-devel openssl-devel dkms mysql-devel libedit-devel postgresql-devel postgresql-server
-                                                sendmail cmake byacc libglvnd-devel mdadm]
+                                                libffi-devel dkms mysql-devel libedit-devel postgresql-devel postgresql-server
+                                                sendmail cmake byacc libglvnd-devel mdadm zlib-devel bzip2 bzip2-devel readline-devel
+                                                sqlite sqlite-devel xz xz-devel findutils]
```
Comment (lukeseawalker) [inline attributes/default.rb]: why these new dependencies? And why only for 'amazon' platform?

File: files/default/attachVolume.py [hunk 4]
```diff
@@ -23,6 +21,20 @@ def convert_dev(dev):
...
         return dev
 
+def get_all_devices():
+    # lsblk -d -n
+    # xvda 202:0    0  17G  0 disk
+    # xvdb 202:16   0  20G  0 disk /shared
+    command = ["/bin/lsblk", "-d", "-n"]
+
+    env = {}
+    env.update(os.environ.copy())
```
Comment (lukeseawalker) [inline files/default/attachVolume.py]: please comment why this is needed

File: files/default/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,4 @@
+awscli
+boto3
+supervisor==3.4.0
```
Comment (lukeseawalker) [inline files/default/requirements.txt]: no need pin supervisor anymore

File: templates/amazon/parallelcluster_supervisord.conf.erb [hunk 6]
```diff
@@ -5,12 +5,12 @@
 <% when 'MasterServer' -%>
```
Comment (lukeseawalker) [inline templates/amazon/parallelcluster_supervisord.conf.erb]: Do we still need this 'amazon' declination? It seems the files are now the same, so we can keep just the default

File: files/default/ec2_dev_2_volid.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/opt/parallelcluster/virtualenv/bin/python
```
Comment (lukeseawalker) [inline files/default/ec2_dev_2_volid.py]: Is it possible to use a variable instead of a static full path here?

Comment (sean-smith) [reply files/default/ec2_dev_2_volid.py]: ok, now uses variable

File: recipes/_ec2_udev_rules.rb [hunk 8]
```diff
@@ -13,7 +13,7 @@
 # OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
 # limitations under the License.
 
-cookbook_file 'ec2-volid.rules' do
+template 'ec2-volid.rules' do
   path '/etc/udev/rules.d/52-ec2-volid.rules'
```
Comment (lukeseawalker) [inline recipes/_ec2_udev_rules.rb]: isn't source attribute missing?

File: recipes/base_install.rb [hunk 9]
```diff
@@ -202,7 +141,7 @@
 end
 
 # Put init script in place
-cookbook_file "supervisord-init" do
+template "supervisord-init" do
   path "/etc/init.d/supervisord"
```
Comment (lukeseawalker) [inline recipes/base_install.rb]: missing source attribute

Comment (sean-smith) [reply recipes/base_install.rb]: added

**Comment by sean-smith** [comment]
Needs accompanying https://github.com/aws/aws-parallelcluster-node/pull/159 patch

File: files/default/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,5 @@
+awscli
+boto3
+supervisor
+six
```
Comment (lukeseawalker) [inline files/default/requirements.txt]: remove six, it's a -node dep

**Pull request merged**